feat: ID-4134: Support Bootstrap Flow for Wallet Initial Transaction

naveen-imtb opened the pull request:
**Summary:**
This PR implements support for the bootstrap flow in the v1 wallet system. This is done by checking if the initial txn is signed by the Immutable signer without requiring a pre-stored image hash.

**Motivation:**

**Pre-change:**
When a new wallet is created, the image hash (representing the wallet's signer configuration) needs to be stored before transactions can be validated. In the absence of the imageHash, the first transaction is validated by recalculating the CREATE2 address of the wallet using the imageHash as the salt. The imageHash is retrieved from the signed message and the message payload. If the calculated address (CFA) matches the deployed wallet address then the imageHash is stored and used to validate subsequent transactions.

**The problem**
Currently in the Immutable ZKEVM chain, the primary wallet owner / signer identity instrument is stored by a 3rd party and we access the signer via their TEE. In a multi-chain world, we would potentially be integrating with different infrastructure providers across chain and therefore have different signers. 
The primary wallet owner / signer influences the CFA i.e the user's Passport wallet address and since we need to preserve the same CFA, the current model of validating the first transaction by verifying if the recalculated wallet CFA matches the deployed wallet address doesn't work.

**Post-change**
To work around the above problem, this PR introduces the following changes:
- A new wallet's first transaction is submitted with a signature from the Immutable Signer contract
- During signature validation, the contract checks:
  - If the current nonce is 1 (meaning this is the first transaction, since nonce was 0 before increment)
  - If the Immutable Signer contract address is among the signers in the signature
  - If both conditions are met, the signature is automatically approved and the computed image hash is stored
- Subsequent transactions follow the standard validation path, checking against the stored image hash

Overall Flow:
<img width="6244" height="4206" alt="Passport - Initial Txn-2025-12-04-043344" src="https://github.com/user-attachments/assets/ba387fe3-2af9-41da-b864-bb72f777530d" />

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces immutable signer–based bootstrap to validate the first wallet transaction without user signature, updating constructors, deployments, and compiler settings.
> 
> - **Contracts**:
>   - **`ModuleAuthDynamic`**: Adds `IMMUTABLE_SIGNER_CONTRACT` and overrides signature validation to auto-approve first tx (nonce=0) when immutable signer is present; integrates nonce/image checks via `ModuleStorage`/`NonceKey`.
>   - **`MainModuleDynamicAuth` + mocks (`MainModuleMockV1/2/3`)**: Constructors now accept and pass `_immutableSignerContract`.
>   - **`ModuleAuth`**: Exposes `FLAG_*` as `internal`; marks validation functions `virtual` to enable overrides.
> - **Deployment**:
>   - **`scripts/deploy.ts` & `scripts/step4.ts`**: Deploy `ImmutableSigner` before `MainModuleDynamicAuth` and supply its address during deployment; output includes immutable signer details.
> - **Build/Config**:
>   - **`hardhat.config.ts`**: Enables `viaIR` in Solidity compiler settings.
>   - **`.gitignore`**: Ignores `lib/`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 7716f6532530d5aba56b1f75c2dcc96b9c9c95ca. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

naveen-imtb committed:
chore: ID-4134: bootstrap wallet for initial txn

Hook into signature validation to bootstrap wallet deployment for first txn without user signature.

naveen-imtb committed:
fix: update nonce value for first txn

drinkcoffee commented on src/contracts/modules/commons/ModuleAuthDynamic.sol:
```diff
@@ -4,19 +4,156 @@ pragma solidity 0.8.17;
 
     INIT_CODE_HASH = initCodeHash;
     FACTORY = _factory;
+    IMMUTABLE_SIGNER_CONTRACT = _immutableSignerContract;
```
Add a check that _immutableSignerContract is non-zero

naveen-imtb replied:
Added a require statement at the entry point of the constructor.

drinkcoffee commented on src/contracts/modules/commons/ModuleAuthDynamic.sol:
```diff
@@ -4,19 +4,156 @@ pragma solidity 0.8.17;
+      }
+
+      // Defensive check: compare extracted address with target address
+      if (IMMUTABLE_SIGNER_CONTRACT != address(0) && addr == IMMUTABLE_SIGNER_CONTRACT) {
```
Remove check that Immtuable signer contract is zero

naveen-imtb commented on hardhat.config.ts:
```diff
@@ -16,16 +16,19 @@ loadAndValidateEnvironment();
+    compilers: [{
+      version: '0.8.17',
+      settings: {
+        viaIR: true,
```
@drinkcoffee I tweaked this settling to resolve a "stack too deep" error. Not sure if there's any unintended side effects that I'm not aware of. Pls advise.

naveen-imtb replied:
@drinkcoffee I ended up changing this because it caused issues with the existing tests where revert messages were validated. I had to introduce an inline struct to reduce the number of variables on the stack.

naveen-imtb committed:
chore: ID-4134: fix review comments

naveen-imtb committed:
chore: ID-4134: add tests

naveen-imtb committed:
chore: remove usage of viaIR

remove usage of viaIR & update logic to workaround deep stack issue

naveen-imtb committed:
chore: ID-4134: threat model

naveen-imtb committed:
chore: ID-4134: update tests.

naveen-imtb committed:
chore: ID-4134: update threat model

drinkcoffee commented on src/contracts/modules/commons/ModuleAuthDynamic.sol:
```diff
@@ -4,19 +4,163 @@ pragma solidity 0.8.17;
+   *        uint8 01,  uint8 weight_5, address signer_5
+   *      )
+   */
+  function _signatureValidation(
```
This function appears to be the same as the one in ModuleAuth. Is it needed?

naveen-imtb replied:
Addressed.

drinkcoffee commented on src/contracts/modules/commons/ModuleAuthDynamic.sol:
```diff
@@ -4,19 +4,163 @@ pragma solidity 0.8.17;
+      uint256 rindex     // read index
+    ) = _signature.readFirstUint16();
+
+    SignatureValidationState memory state = SignatureValidationState({
```
Storing these values in memory rather than on the stack will make the code more expensive to run

naveen-imtb replied:
Had to go down this path to workaround the "stack too deep" issue 🤔  Any suggestions on alternate approaches?

drinkcoffee replied:
Use local variables and change the FLAG_DYNAMIC_SIGNATURE block to:
        // Read signer
        (addr, rindex) = _signature.readAddress(rindex);

        bytes memory signature;
        {
          // Read signature size
          uint256 size;
          (size, rindex) = _signature.readUint16(rindex);

          // Read dynamic size signature
          (signature, rindex) = _signature.readBytes(rindex, size);
        }
        require(isValidSignature(_hash, addr, signature), "ModuleAuthDynamic#_signatureValidation: INVALID_SIGNATURE");


Doing this helps the compiler by providing scope to size.

naveen-imtb replied:
I wasn't able to resolve the stack issue with block scoping, have retained the memory variable approach for now.

drinkcoffee replied:
When I retried, I couldn't get that specific fix to work. I am not sure why.

drinkcoffee replied:
I have created a PR which takes an alternative approach to allow the variables to be on the stack rather than in memory. Approve it and merge it into this one if you are happy with the approach.
https://github.com/immutable/wallet-contracts/pull/76

drinkcoffee commented on src/contracts/modules/commons/ModuleAuthDynamic.sol:
```diff
@@ -4,19 +4,163 @@ pragma solidity 0.8.17;
+      return (true, true, state.imageHash);
+    }
+
+    (bool verified, bool needsUpdate) = _isValidImage(state.imageHash);
```
The CFA will match the calculated image hash: a combination of the user key's address and the Immutable key's address.  When an alternative user secret provider is used, the user's key will be different, so the address will be different, and hence the calculated image hash and then CFA will be different. 
If this is the case, then the _isValidImage check will fail here: https://github.com/immutable/wallet-contracts/blob/main/src/contracts/modules/commons/ModuleAuthDynamic.sol#L35

drinkcoffee replied:
What am I missing in my understanding?

naveen-imtb replied:
In the scenario where a user's CFA was generated with signer A but the user is now associated with signer B (e.g., due to using different identity providers on different chains), the `_isValidImage` check will not be invoked on first transaction.
Instead, during the first transaction (when currentNonce == 1), the contract performs an early exit after validating that the Immutable Signer is present among the signers. 
At [line 158-160](https://github.com/immutable/wallet-contracts/blob/277c76d9d7b56075e8482801380b0e0b68a08f19/src/contracts/modules/commons/ModuleAuthDynamic.sol#L158):
```
if (currentNonce == 1 && state.immutableSignerContractFound) {  
    return (true, true, state.imageHash);
}
```
This bootstrap flow bypasses the `_isValidImage` check entirely. The computed image hash (derived from the new signer configuration, including signer B) is returned and subsequently stored via `updateImageHashInternal()`. 
This is the intentional design of the bootstrap flow - it allows the Immutable Signer to initialize the wallet with a different user key configuration while preserving the original CFA.
The `_isValidImage` check at [line 162](https://github.com/immutable/wallet-contracts/blob/277c76d9d7b56075e8482801380b0e0b68a08f19/src/contracts/modules/commons/ModuleAuthDynamic.sol#L162) is only reached for subsequent transactions (nonce ≥ 2), where it validates against the already-stored image hash.

drinkcoffee replied:
That makes sense. Thank you for the detailed explanation.

JCSanPedro commented on audits/202412-threat-model-bootstrap-flow-v1.md:
We validate that the `immutableSignerContract` address is non-zero, but should we also validate that the actual [primarySigner](https://github.com/immutable/wallet-contracts/blob/277c76d9d7b56075e8482801380b0e0b68a08f19/src/contracts/signer/ImmutableSigner.sol#L22) address value is non-zero as well?

naveen-imtb replied:
Yes probably should have but unfortunately the signer contract is non upgradable. We however do have the ability to rotate the signer EOA.

JCSanPedro commented on audits/202412-threat-model-bootstrap-flow-v1.md:
Oh do we already have key rotation in place for the Immutable Signer private key?

naveen-imtb replied:
Yes we do and we also support a rollover period.

JCSanPedro commented on audits/202412-threat-model-bootstrap-flow-v1.md:
I don't think monitoring `PrimarySignerUpdated` would allow us to detect if the `ImmutableSigner` was compromised? As the attacker wouldn't update the signer in this case.

Though this would allow us to detect if the [SIGNER_ADMIN](https://github.com/immutable/wallet-contracts/blob/277c76d9d7b56075e8482801380b0e0b68a08f19/src/contracts/signer/ImmutableSigner.sol#L39) key was compromised.

naveen-imtb replied:
You're right, the private key of the Immutable signer being compromised will not be detected via the suggested monitoring. I'll fix it up.

JCSanPedro commented on tests/ModuleAuthDynamic.spec.ts:
```diff
@@ -0,0 +1,682 @@
+    })
+  })
+})
+
```
I think we can add tests to validate the bootstrap flow (and subsequent transactions) works if the immutable signer is updated (and doesn't work using the old signer address).

naveen-imtb replied:
I don't think there is value in testing the signer key rotation as part of this work. The bootstrap flow doesn't actually validate the signature, it only extracts the signer information from the signature and attempts to match it to the Immutable signer contract address. The contract address itself is fixed and not affected by signer rotation.

shineli1984 commented on audits/202412-threat-model-bootstrap-flow-v1.md:
key rotation needs startup implementation contract to point to a new version?

naveen-imtb replied:
The Immutable signer contract's primary signer can be rotated without impacting the CFA address or other wallet workflow. The implementation contract only has the Immutable Signer's contract address persisted in storage so key rotation doesn't require any update to the implementation contract either.

naveen-imtb committed:
chore: remove redundant function override

naveen-imtb committed:
chore: review comments.

naveen-imtb committed:
fix: use blockchain timestamp instead of system time in expiration tests

Eliminate the clock skew issue between the JavaScript runtime and the blockchain using blockchain timestamp

drinkcoffee commented on src/contracts/modules/commons/ModuleAuthDynamic.sol:
```diff
@@ -4,19 +4,125 @@ pragma solidity 0.8.17;
+      (flag, addrWeight, state.rindex) = _signature.readUint8Uint8(state.rindex);
+
+      if (flag == FLAG_ADDRESS) {
+        // Read plain address
```
I think we only use FLAG_SIGNATURE and FLAG_DYNAMIC_SIGNATURE. If I am correct, then we should remove the FLAG_ADDRESS option.

At attacker could create two valid signatures that are FLAG_SIGNATURE, and include the Immutable signer as a FLAG_ADDRESS. The checks below would deem the signature valid and bootstrap the wallet at line 109 and 120.

drinkcoffee replied:
They don't need to create two valid signatures, as we don't check the weight. Hence, they could just create a _signature block that had Immutable signer as a FLAG_ADDRESS, and then they could set the image hash to a value of their choosing.

naveen-imtb replied:
You're right @drinkcoffee 

This is a critical bug. I've updated the contract with the following changes:
- Removed FLAG_ADDRESS support - This prevents attackers from including the immutable signer address in the signature without actually providing a valid signature from that address.
- Required verified signatures from the Immutable signer contract - The immutable signer must now actually sign the transaction (via FLAG_SIGNATURE or FLAG_DYNAMIC_SIGNATURE) rather than just being listed as an address. The immutableSignerContractSigned flag is only set to true after the signature has been cryptographically verified.

These changes ensure that the bootstrap flow can only be triggered when the Immutable signer contract has genuinely signed the transaction, preventing unauthorized bootstrap attacks.

w3njah replied:
@naveen-imtb @drinkcoffee Not sure if it's intentional or not, but removing `FLAG_ADDRESS` completely creates the following side effect: it's no longer possible to configure SCW to accept `N` out of `M` signatures.

The whole point of `FLAG_ADDRESS` is to be able to specify signer's address without providing its signature, so that calculated `imageHash` would match the stored one and the validity of the total signature would be determined by the weight > threshold condition.

It might be fine for us if we are 100% certain we always want "2 out of 2" configuration for our SCWs

naveen-imtb committed:
chore: fix FLAG_ADDRESS security vulnerability

Fix bootstrap signature validation to require verified signature from Immutable signer and meet weight threshold, removing FLAG_ADDRESS support to prevent unauthorized wallet takeover on first transaction.

w3njah commented on audits/202412-threat-model-bootstrap-flow-v1.md:
```diff
@@ -0,0 +1,513 @@
+
+**PR Reference:** [#74 - feat: ID-4134: Support Bootstrap Flow for Wallet Initial Transaction](https://github.com/immutable/wallet-contracts/pull/74)
+
+**Date:** December 2024
```
```suggestion
**Date:** December 2025
```

w3njah commented on audits/202412-threat-model-bootstrap-flow-v1.md:
```diff
@@ -0,0 +1,513 @@
+
+When a new wallet is created in the existing system, the image hash (representing the wallet's signer configuration) needs to be stored before transactions can be validated. In the absence of a stored imageHash, the first transaction is validated by:
+
+1. Recalculating the CREATE2 address of the wallet using the imageHash as the salt
+2. Retrieving the imageHash from the signed message and payload
```
```suggestion
1. Retrieving the imageHash from the signed message and payload
2. Recalculating the CREATE2 address of the wallet using the imageHash as the salt
```
